File: examples/go/rebalancing_schedule_resource_example.go
package main

import (
	"encoding/json"

	"github.com/castai/pulumi-castai/sdk/go/castai" // Assuming main package
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// runRebalancingScheduleResourceExample demonstrates the schedule resource.
// Renamed from main to avoid conflicts.
func runRebalancingScheduleResourceExample(ctx *pulumi.Context) error {

	selectorJSON, _ := json.Marshal(map[string]interface{}{
		"nodeSelectorTerms": []map[string]interface{}{{
			"matchExpressions": []map[string]interface{}{{
				"key":      "scheduling.cast.ai/spot",
				"operator": "Exists",
			}},
		}},
	})

	spotSchedule, err := castai.NewRebalancingSchedule(ctx, "example-go-spot-schedule", &castai.RebalancingScheduleArgs{
		Name: pulumi.String("rebalance-go-spots-hourly"),
		Schedule: &castai.RebalancingScheduleScheduleArgs{
			Cron: pulumi.String("0 * * * *"), // Every hour at minute 0
		},
		TriggerConditions: &castai.RebalancingScheduleTriggerConditionsArgs{
			SavingsPercentage: pulumi.Int(15), // Trigger if potential savings > 15%
		},
		LaunchConfiguration: &castai.RebalancingScheduleLaunchConfigurationArgs{
			NodeTtlSeconds:   pulumi.Int(600), // Only consider nodes older than 10 minutes
			NumTargetedNodes: pulumi.Int(5),   // Target up to 5 nodes per run
			// RebalancingMinNodes:    pulumi.Int(2),    // Optional: Minimum nodes to keep
			// KeepDrainTimeoutNodes: pulumi.Bool(false), // Optional
			Selector: pulumi.String(string(selectorJSON)), // Node selector as a JSON string
			ExecutionConditions: &castai.RebalancingScheduleLaunchConfigurationExecutionConditionsArgs{
				Enabled: pulumi.Bool(true),
				// AchievedSavingsPercentage: pulumi.Int(5), // Optional: Only execute if savings > 5%
			},
		},
	})
	if err != nil {
		return err
	}

	// Export the schedule ID
	ctx.Export("spot_schedule_go_id", spotSchedule.ID())

	return nil
}

/*
// Example of how you might call this from your main function:
func main() {
    pulumi.Run(func(ctx *pulumi.Context) error {
        if err := runRebalancingScheduleResourceExample(ctx); err != nil {
            return err
        }
        // ... potentially call other example functions ...
        return nil
    })
}
*/

File: examples/python/rebalancing_schedule_resource_example.py
"""
CAST AI Rebalancing Schedule Example for Python

This example demonstrates how to create a rebalancing schedule in CAST AI.
"""

import pulumi
import pulumi_castai as castai
import json
import os
from typing import Dict, List, Optional, Any
from pydantic import BaseModel, Field

# Import the CAST AI provider and models
from models import Provider as ProviderModel

# Initialize the CAST AI provider
api_token = os.environ.get("CASTAI_API_TOKEN", "your-api-token-here")
provider_config = ProviderModel(api_token=api_token)
provider = castai.Provider("castai-provider",
    api_token=provider_config.api_token,
    api_url=provider_config.api_url
)

# Pydantic models for Rebalancing Schedule
class ScheduleArgs(BaseModel):
    """Configuration for the schedule's cron expression."""
    cron: str

class TriggerConditionsArgs(BaseModel):
    """Conditions that trigger the rebalancing."""
    savings_percentage: int

class ExecutionConditionsArgs(BaseModel):
    """Conditions for execution of the rebalancing."""
    enabled: bool
    achieved_savings_percentage: Optional[int] = None

class LaunchConfigurationArgs(BaseModel):
    """Configuration for the rebalancing launch."""
    node_ttl_seconds: int
    num_targeted_nodes: int
    rebalancing_min_nodes: Optional[int] = None
    keep_drain_timeout_nodes: Optional[bool] = None
    selector: str  # JSON string
    execution_conditions: ExecutionConditionsArgs

class RebalancingScheduleModel(BaseModel):
    """Full configuration for a rebalancing schedule."""
    name: str
    schedule: ScheduleArgs
    trigger_conditions: TriggerConditionsArgs
    launch_configuration: LaunchConfigurationArgs

# Create the rebalancing schedule model
spot_schedule_model = RebalancingScheduleModel(
    name="rebalance-py-spots-hourly",
    schedule=ScheduleArgs(
        cron="0 * * * *",  # Every hour at minute 0
    ),
    trigger_conditions=TriggerConditionsArgs(
        savings_percentage=15,  # Trigger if potential savings > 15%
    ),
    launch_configuration=LaunchConfigurationArgs(
        node_ttl_seconds=600,  # Only consider nodes older than 10 minutes
        num_targeted_nodes=5,  # Target up to 5 nodes per run
        # rebalancing_min_nodes=2,  # Optional: Minimum nodes to keep
        # keep_drain_timeout_nodes=False,  # Optional
        selector=json.dumps({  # Node selector as a JSON string
            "nodeSelectorTerms": [{
                "matchExpressions": [{
                    "key": "scheduling.cast.ai/spot",
                    "operator": "Exists",
                }],
            }],
        }),
        execution_conditions=ExecutionConditionsArgs(
            enabled=True,
            # achieved_savings_percentage=5,  # Optional: Only execute if savings > 5%
        ),
    )
)

# Convert the model to a dictionary
schedule_args = spot_schedule_model.dict(exclude_none=True)

# Create the rebalancing schedule resource
try:
    # Use the direct RebalancingSchedule type
    spot_schedule = castai.RebalancingSchedule(
        "example-py-spot-schedule",
        **schedule_args,
        opts=pulumi.ResourceOptions(provider=provider)
    )
    
    # Export the schedule ID
    pulumi.export("spot_schedule_py_id", spot_schedule.id)
except Exception as e:
    pulumi.log.warn(f"Failed to create rebalancing schedule: {str(e)}")
    # Export the configuration as demonstration
    pulumi.export("rebalancing_schedule_config", schedule_args)
    pulumi.export("provider_configured", True) 